buku.py

```diff
@@ -1740,6 +1740,142 @@ def exportdb(self, filepath, taglist=None):
+
+            bookmark_tags = []
+            for bookmark_tag_id in bookmark_tags_ids:
+                res = cur.execute(
+                    'SELECT title FROM moz_bookmarks WHERE id={}'.format(bookmark_tag_id)
+                )
+                bookmark_tags.append(res.fetchone()[0])
+
+            # get the url
+            res = cur.execute(
+                'SELECT url FROM moz_places WHERE id={}'.format(place_id)
+            )
+            url = res.fetchone()[0]
+
+            # get the title
+            res = cur.execute(
+                'SELECT title FROM moz_bookmarks '
+                'WHERE fk={} AND title!="" LIMIT 1'.format(place_id)
+            )
+            title_data = res.fetchone()
+            if title_data:
+                title = title_data[0]
+            else:
+                title = ''
+            tags = self.parse_tags(bookmark_tags)
+            self.add_rec(url, title, tags)
+        try:
+            cur.close()
+            conn.close()
+        except Exception:
+            print('Error here')
+
+    def import_from_browser(self):
+        '''Import bookmarks from a browser default database file.
+        Supports Firefox and Google Chrome.
+
+        :return: True on success, False on failure
+        '''
+        if sys.platform.startswith('linux'):
+            GC_BM_DB_PATH = '~/.config/google-chrome/Default/Bookmarks'
+            DEFAULT_FF_FOLDER = os.path.expanduser('~/.mozilla/firefox')
+            profile = get_firefox_profile_name(DEFAULT_FF_FOLDER)
+            FF_BM_DB_PATH = (
+                '~/.mozilla/firefox/{}.default/places.sqlite'.format(profile)
+            )
+
+        elif sys.platform == 'darwin':
+            GC_BM_DB_PATH = (
+                '~/Library/Application Support/Google/Chrome/Default/Bookmarks'
+            )
+            DEFAULT_FF_FOLDER = os.path.expanduser('~/Library/Application Support/Firefox')
+            profile = get_firefox_profile_name(DEFAULT_FF_FOLDER)
+
+            FF_BM_DB_PATH = (
+                '~/Library/Application Support/Firefox/{}.default/places.sqlite'.format(profile)
+            )
+
+        elif sys.platform == 'win32':
+            username = os.getlogin()
+            GC_BM_DB_PATH = (
+                'C:/Users/{}/AppData/Local/Google/Chrome/User Data/Default/Bookmarks'.format(username)
+            )
+            DEFAULT_FF_FOLDER = 'C:/Users/{}/AppData/Roaming/Mozilla/Firefox/Profiles'.format(username)
+            profile = get_firefox_profile_name(DEFAULT_FF_FOLDER)
+
+            FF_BM_DB_PATH = (
+                os.path.join(DEFAULT_FF_FOLDER, '{}.default/places.sqlite'.format(profile))
+            )
+
+        else:
+            logerr('Buku does not support {} yet'.format(sys.platform))
+            self.close_quit(1)
+
+        try:
+            webbrowser.get('google-chrome')
+            bookmarks_database = os.path.expanduser(GC_BM_DB_PATH)
+            walk(load_chrome_database(bookmarks_database))
```

Comment:
@alex-bender I can see you are walking through the Chrome json file, But where are the bookmarks actually added? I don't see any `add_rec()` call for Chrome bookmarks.

Reply:
hi @jarun ! I've changed this function to a generator, as `rachmadaniHaryono` proposed, but didn't change the function invocation -- that's my bad. I'll fix it as soon as possible.. Hope it'll be tomorrow -- coz i still didn't recover.

Reply:
I have fixed this in latest trunk. Also made the changes so that auto-import works like importdb().

The only thing that's remaining is:

Import parent folder as tag **for Firefox**.